easy_exceptions/easy_exceptions.py

```diff
@@ -1,14 +1,29 @@
+import threading
+from time import sleep
+
 generated_exceptions = {}
 bound_parent = Exception
+update_mutex = threading.Lock()
+TEST_PARALELLISM = False
 
 def reset_module():
-    global generated_exceptions, bound_parent 
+    # !!!! This function is intended for testing purposes. Using this in production code breaks 
+    # thread safety guarantees
+    global generated_exceptions, bound_parent, update_mutex
     generated_exceptions = {} 
     bound_parent = Exception
+    update_mutex = threading.Lock()
 
 def bind_parent(parent):
     global bound_parent
-    bound_parent = parent
+    acquired = update_mutex.acquire() 
+    if not acquired:
+        raise Exception("Lock acquire while binding parent failed")
```

Comment:
not necessary to change but to make it more easy to to digest, perhaps:
`raise Exception("Failed to acquire lock on GIL while binding parent")`